tests/Functional/App/Controller.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare(strict_types=1);
+
+namespace tests\Libero\ApiProblemBundle\Functional\App;
+
+use Exception;
+use FluentDOM\DOM\Element;
+use Libero\ApiProblemBundle\Exception\ApiProblem;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+
+final class Controller
+{
+    public function success() : Response
+    {
+        return new Response();
+    }
+
+    public function httpException() : Response
+    {
+        throw new HttpException(Response::HTTP_I_AM_A_TEAPOT, 'An HTTP exception');
+    }
+
+    public function exception() : Response
+    {
+        throw new Exception('An exception');
+    }
+
+    public function apiProblem(Request $request) : Response
+    {
+        $apiProblem = new ApiProblem($request);
+        /** @var Element $root */
+        $root = $apiProblem->getDocument()->documentElement;
+
+        $root->appendElement('status', (string) Response::HTTP_TOO_MANY_REQUESTS);
+        $root->appendElement('title', 'Too Many Requests');
+        $root->appendElement('details', 'Calm it down.');
```

Comment:
Shame `DOMDocument` is mutable, but I think there should be a way to add more information in a separate listener (use case: adding the stacktrace in debug mode).

Other approach is to create a custom event rather than using `kernel.exception`. The document's still mutable, but it's then separate from the exception.

Reply:
So there is this exception that travels between listeners, while containing a mutable error document; the contract between the listeners and that shared data structure is loose.

May also want to add information to that error document without having the exception e.g. adding a correlation id at the start of the request or some other debug information in a fingers-crossed mode in case it's needed. This points towards the second solution of a custom event perhaps, but I'm not sure how difficult it is to use.

Reply:
It would be basically the same, just listening for a different event.